src/Uno.UI.RuntimeTests.Engine.Library/UnitTestsUIContentHelper.cs

```diff
@@ -1,207 +1,184 @@
+	}
+
+	public static void RestoreOriginalContent()
+	{
+		if (_originalWindowContent != null && CurrentTestWindow is not null)
+		{
+			CurrentTestWindow.Content = _originalWindowContent;
+			_originalWindowContent = null;
+		}
+	}
+
+	public static Window? CurrentTestWindow
+	{
+		get => _currentTestWindow ?? throw new InvalidOperationException("Current test window not set.");
+		set => _currentTestWindow = value;
+	}
+
+	public static UIElement? Content
+	{
+		get => UseActualWindowRoot && CurrentTestWindow is not null
+			? CurrentTestWindow.Content
+			: EmbeddedTestRoot.GetContent?.Invoke();
+
+		internal set
+		{
+			if (UseActualWindowRoot && CurrentTestWindow is not null)
+			{
+				CurrentTestWindow.Content = value;
+			}
+			else if (EmbeddedTestRoot.SetContent is { } setter)
+			{
+				setter(value);
+			}
+			else
+			{
+				Console.WriteLine("Failed to get test content control");
+			}
+		}
+	}
+
+
+	/// <summary>
+	/// Waits for the dispatcher to finish processing pending requests
+	/// </summary>
+	/// <returns></returns>
+	public static async Task WaitForIdle()
+	{
+		await RootElementDispatcher.RunAsync(UnitTestDispatcherCompat.Priority.Low, () => { });
+		await RootElementDispatcher.RunAsync(UnitTestDispatcherCompat.Priority.Low, () => { });
+	}
+
+	/// <summary>
+	/// Waits for <paramref name="element"/> to be loaded and measured in the visual tree.
+	/// </summary>
+	/// <remarks>
+	/// On UWP, <see cref="WaitForIdle"/> may not always wait long enough for the control to be properly measured.
+	///
+	/// This method assumes that the control will have a non-zero size once loaded, so it's not appropriate for elements that are
+	/// collapsed, empty, etc.
+	/// </remarks>
+	public static async Task WaitForLoaded(FrameworkElement element)
+	{
+		async Task Do()
+		{
+			bool IsLoaded()
+			{
+				if (element.ActualHeight == 0 || element.ActualWidth == 0)
 				{
-					await Do();
+					return false;
 				}
-				else
+
+				if (element is ListView listView && listView.Items.Count > 0 && listView.ContainerFromIndex(0) == null)
 				{
-					TaskCompletionSource<bool> cts = new();
-
-					_ = element.DispatcherQueue.TryEnqueue(DispatcherQueuePriority.Normal, () =>
-					{
-						try
-						{
-
-							cts.TrySetResult(true);
-						}
-						catch (Exception e)
-						{
-							cts.TrySetException(e);
-						}
-					});
-
-					await cts.Task;
+					// If it's a ListView, wait for items to be populated
+					return false;
 				}
+
+				return element.IsLoaded;
+			}
+
+			await WaitFor(IsLoaded, message: $"{element} loaded");
+		}
+#if __WASM__ // Adjust for re-layout failures in When_Inline_Items_SelectedIndex, When_Observable_ItemsSource_And_Added, When_Presenter_Doesnt_Take_Up_All_Space
+		await Do();
+#else
+		var dispatcher = UnitTestDispatcherCompat.From(element);
+		if (dispatcher.HasThreadAccess)
+		{
+			await Do();
+		}
+		else
+		{
+			await dispatcher.RunAsync(UnitTestDispatcherCompat.Priority.Normal, () => { });
```

Comment:
not sure if it is intended that this block here just await a "do nothing" on ui thread?
even in uno too: https://github.com/unoplatform/uno/blob/4.6.19/src/Uno.UI.RuntimeTests/IntegrationTests/common/TestServices.WindowHelper.cs#L154-L169

Reply:
shouldnt we call `Do()` here?

Reply:
Seems so 👀

Reply:
@jeromelaban

Reply:
This is a weird helper indeed. It should await `Do()` as well.